python/lsst/daf/butler/datastores/fileLikeDatastore.py

```diff
@@ -595,6 +639,69 @@ def getUri(self, ref, predict=False):
 
         return location.uri
 
+    @transactional
+    def trash(self, ref):
+        """Indicate to the Datastore that a Dataset can be removed.
+
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            Reference to the required Dataset.
+
+        Raises
+        ------
+        FileNotFoundError
+            Attempt to remove a dataset that does not exist.
+        """
+        # Get file metadata and internal metadata
+        log.debug("Trashing %s in datastore %s", ref, self.name)
+        location, _ = self._get_dataset_location_info(ref)
+        if location is None:
+            raise FileNotFoundError(f"Requested dataset ({ref}) does not exist")
+
+        if not self._artifact_exists(location):
+            raise FileNotFoundError(f"No such artifact: {location.uri}")
+
+        # Mark dataset as trashed
+        self._move_to_trash_in_registry(ref)
+
+    @transactional
+    def emptyTrash(self):
+        """Remove all datasets from the trash.
+        """
+        log.debug("Emptying trash in datastore %s", self.name)
+        trashed = self.registry.getTrashedDatasets(self.name)
+        artifactsToRemove = set()
+
+        for ref in trashed:
+            location, _ = self._get_dataset_location_info(ref)
+
+            if location is None:
+                raise FileNotFoundError(f"Requested dataset ({ref}) does not exist")
+
+            if not self._artifact_exists(location):
+                raise FileNotFoundError(f"No such file: {location.uri}")
```

Comment:
I suspect making this a silent no-op will make concurrent calls to `emptyTrash` safer, and maybe totally safe.

Reply:
Ok. Concurrent access would suggest that we should also skip the earlier test since it implies that the trashed file has also been removed from the internal registry as well as having the file removed.